packages/labs/context/src/lib/decorators/provide.ts

```diff
@@ -76,3 +71,46 @@ export function provide<ValueType>({
     },
   });
 }
+
+type ProvideDecorator<ContextType> = {
+  <K extends PropertyKey, Proto extends ReactiveElement>(
+    protoOrDescriptor: Proto,
+    name?: K
+    // Note TypeScript requires the return type to be `void|any`
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ): FieldMustMatchContextType<Proto, K, ContextType, void | any>;
+};
+
+type FieldMustMatchContextType<
+  Obj,
+  Key extends PropertyKey,
+  ContextType,
+  ReturnValue
+> =
+  // First we check whether the object has the property as a required field
+  Obj extends Record<Key, infer ProvidingType>
+    ? // Ok, it does, just check whether it's ok to assign the
+      // provided type to the consuming field
+      [ProvidingType] extends [ContextType]
+      ? ReturnValue
+      : {
+          message: 'providing field not assignable to context';
+          context: ContextType;
+          provided: ProvidingType;
+        }
+    : // Next we check whether the object has the property as an optional field
+    Obj extends Partial<Record<Key, infer Providing>>
+    ? // Check assignability again. Note that we have to include undefined
+      // here on the consuming type because it's optional.
```

Comment:
Nit: Should this comment reference the `providing` type?

Reply:
Good catch!